docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,242 @@
+
+An optional field is proposed for users to add to their Infrastructure Machine Templates which will
+carry information about the CPU, memory, and GPU requirements for the machines created from the template.
+The cluster autoscaler will then utilize this information by reading the appropriate infrastructure
+reference from the resource it is scaling (MachineSet or MachineDeployment).
+
+A user may override the field in the associated infrastructure template by applying annotations to the
+MachineSet or MachineDeployment in question. In these cases the autoscaler will evaluate the annotation
+information in favor of reading the information from the status.
+
+### User Stories
+
+#### Story 1
+
+As a Cluster API user, I would like to reduce my operating costs by scaling down my workload
+cluster when they are not in use. Using the cluster autoscaler with a minimum size of zero for
+a MachineSet or MachineDeployment will allow me to automate the scale down actions for my clusters.
+
+#### Story 2
+
+As a Cluster API user, I would like to have special resource nodes (eg GPU enabled) provided when needed by workloads
+without the need for human intervention. As these nodes might be more expensive, I would also like to return them when
+not in use. By using the cluster autoscaler with a zero-sized MachineSet or MachineDeployment, I can automate the
+creation of nodes that will not consume resources until they are required by applications on my cluster.
+
+### Implementation Details/Notes/Constraints
+
+Cloud providers, or users, can add an optional field to their machine template infrastructure objects to indicate
+the resources provided by the machine. Internally, this field will be represented by a Go `map` type
+utilizing named constants for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values
+(similar to how resource limits and requests are handled for pods).
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// DockerMachineTemplateSpec defines the desired state of DockerMachineTemplate.
+type DockerMachineTemplateSpec struct {
+    AutoscalerCapacity ResourceList `json:"autoscalerCapacity,omitempty"`
+    Template DockerMachineTemplateResource `json:"template"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
+
+When used as a manifest, it would look like this:
+
+```
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: DockerMachineTemplate
+metadata:
+  name: workload-md-0
+  namespace: default
+spec:
+  autoscalerCapacity:
+    memory: 500mb
+    cpu: "1"
+    gpu: "1"
+  template:
+    spec: {}
+```
+
+If a user wishes to override the resources hints as provided from the machine template, they may
```

Comment:
just curious what does `hints` refer to?

Reply:
that is the previous name i had been using, i need to update that. thanks!